utils/wasm-gen/src/utils.rs

```diff
@@ -0,0 +1,212 @@
+// This file is part of Gear.
+
+// Copyright (C) 2022 Gear Technologies Inc.
+// SPDX-License-Identifier: GPL-3.0-or-later WITH Classpath-exception-2.0
+
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+use gear_wasm_instrument::parity_wasm::{
+    builder,
+    elements::{self, FuncBody, ImportCountType, Instruction, Module, Type, ValueType},
+};
+use std::collections::{HashMap, HashSet};
+
+const PREALLOCATE: usize = 1_000;
+
+enum Color {
+    Grey,
+    Black,
+}
+
+/// Find possible call recursions in `module`. Calls `callback` with
+/// functions indexes chain and an index creating a recursion.
+///
+/// Used algorithm is based on Depth-First Search (DFS) algorithm for
+/// loops detection in graphs.
+pub fn find_recursion<Callback>(module: &Module, mut callback: Callback)
+where
+    Callback: FnMut(&[usize], usize),
+{
+    let function_bodies = match module.code_section() {
+        Some(s) if !s.bodies().is_empty() => s.bodies(),
+        _ => return,
+    };
+
+    let import_count = module.import_count(ImportCountType::Function);
+
+    let mut colored_list = Vec::<HashMap<_, _>>::with_capacity(function_bodies.len());
```

Comment:
Is this type annotation really needed?
```suggestion
    let mut colored_list = Vec::with_capacity(function_bodies.len());
```